geode-core/src/main/java/org/apache/geode/internal/InternalDataSerializer.java

```diff
@@ -2480,7 +2481,7 @@ public static void invokeFromData(Object ds, DataInput in)
           ((DataSerializable) ds).fromData(in);
         }
       }
-    } catch (EOFException | ClassNotFoundException | CacheClosedException ex) {
+    } catch (EOFException | ClassNotFoundException | CacheClosedException | SocketException ex) {
```

Comment:
We did consider a variation of what you are suggesting, but fundamentally we saw a SocketException as an indication that the connection to the client was in a bad state and figured we should treat it the same as an EOFException.

In the context of a gateway sender, the SerializationException that was wrapping the SocketException would cause all event processing to stop, which is undesirable.  There is handling for all IOExceptions (including SocketExceptions) in the gateway sender, so that it attempts to connect to a different locator instead of shutting down the event processor.  As you suggested, we could add handling at that part of the code to catch SerializationExceptions, check if the inner exception is a SocketException, and retry.  Do you prefer that options?  I assume you may prefer it because the scope of the change is minimized.  

Again, our thought for the change in this PR is that a SocketException is similar to the EOFException in that it indicates a bad connection, and callers have their own handling for these exception types (i.e. retrying in the gateway sender case).  We spent some time looking at various callers and didn't see any red flags for this change.  I could be convinced to instead make the change at the gateway sender logic if there is fear of undesired side effects though.  Let me know.

Reply:
If you look at: InternalDataSerializer.readDataSerializable(DataInput) you will see it calls your modified invokeFromData but turns the SocketException into a SerializationException. So my only  concern is that you will need to change other places to treat SocketException like an EOFException.  Same goes for readDataSerializableFixedID except it currently also transforms EOFException. invokeFromData has lots of other callers but I don't know if any of them have a catch that rethrows. I'm okay with you treating a SocketException like EOFException just wanted to make sure you covered the other code paths and thought it might be easier to handle the SerializationException. But I'll defer to you to pick which is best.

Reply:
Is there a reason not to catch `IOException` more generally?